SPEC.md

```diff
@@ -10164,6 +10112,84 @@ Example output:
 </p>
 </details>
 
+### `length`
+
+```
+Int length(Array[X]|Map[X, Y]|Object|Struct|String)
+```
+
+Returns the length of the input argument as an `Int`:
+
+* For an `Array[X]` argument: the number of elements in the array.
+* For a `Map[X, Y]` argument: the number of items in the map.
+* For a `Struct` or `Object` argument: the number of members in the collection.
+* For a `String` argument: the number of characters in the string.
+
+Note that every struct of a given type will always have the same length, equal to the number of members in the struct definition, regardless of whether any of the members are optional.
+
+**Parameters**
+
+1. `Array[X]`|`Map[X, Y]`|`Object`|`Struct`|`String`: A collection or string whose elements are to be counted.
+
+**Returns**: The length of the collection/string as an `Int`.
+
+<details>
+<summary>
+Example: test_length.wdl
+
+```wdl
+version 1.2
+
+struct Name {
+  String first
+  String last
+  String? middle
+}
+
+workflow test_length {
+  Array[Int] xs = [1, 2, 3]
+  Array[String] ys = ["a", "b", "c"]
+  Array[String] zs = []
+  Map[String, Int] m = {"a": 1, "b", 2}
+  String s = "ABCDE"
+  Name name = Name {
+    first: "John",
+    last: "Doe"
+  }
+
+  output {
+    Int xlen = length(xs)
+    Int ylen = length(ys)
+    Int zlen = length(zs)
+    Int mlen = length(m)
+    Int slen = length(s)
+    Int name_len = length(name)
+  }
+}
+```
+</summary>
+<p>
+Example input:
+
+```json
+{}
+```
+
+Example output:
+
+```json
+{
+  "test_length.xlen": 3,
+  "test_length.ylen": 3,
+  "test_length.zlen": 0,
+  "test_length.mlen": 2,
+  "test_length.slen": 2,
```

Comment:
Shouldn't `test_length.slen` be `5` and `test_length.name_len` be `2`?

Reply:
+1

Reply:
Yes `slen` should be `5`. For `name_length`, since `name` is a struct instance, all members are always present; an optional member whose value is not initialized is set to `None`.

I'm not sure if allowing struct arguments to `contains_key` and `length` is actually useful. I thought it made sense to add it since `Object` is supported, but maybe it just adds more confusion than it's worth.

Reply:
I think it might just be more confusing if all arguments are always present.

Reply:
Done - removed `Struct` as an allowed argument